### PR TITLE
MM-16946 SubMenu in post dropdown menu 

### PR DESCRIPTION
**Pull request by scottleedavis**
###### Overview

```javascript
        const rootRegisterMenuItem = registry.registerPostDropdownMenuItem(
            <FormattedMessage
                id='submenu.menu'
                defaultMessage='Submenu Example'
            />
        );

        const firstItem = (
            <FormattedMessage
                id='submenu.first'
                key='submenu.first'
                defaultMessage='First Item'
            />
        );
        const sub1RegisterMenuItem = rootRegisterMenuItem(
            firstItem,
            () => {
                store.dispatch(postDropdownSubMenuAction(firstItem));
            }
        );
```

Adds an example submenu using registerPostDropdownMenuItem.

img

###### Related PR
* https://github.com/mattermost/mattermost-webapp/pull/3357

### CODE REVIEW

**Comment by hanzei** [comment]
No problem Scott. They will get squashed on merge.

**Comment by scottleedavis** [comment]
Made an update that fixes a bug @saturninoabril found in the PR for webapp.  Perhaps it solves the issue for your too @prapti ?

**Comment by prapti** [comment]
Cool. I'll give it a try :) Thanks for the update @scottleedavis!

File: webapp/i18n/es.json [hunk 1]
```diff
@@ -10,5 +10,9 @@
   "sidebar.enabled": "Habilitado",
   "tooltip.message": "Esta es una información sobre herramientas personalizada del complemento de demostración",
   "useractions.action": "Acción",
-  "useractions.demo": "Complemento de Demostración: "
+  "useractions.demo": "Complemento de Demostración: ",
+  "submenu.menu": "Ejemplo de submenú",
+  "submenu.first": "Primer iteam",
```
Comment (jespino) [inline webapp/i18n/es.json]: `item` not `iteam` but probably it should sound more natural in spanish like `elemento`.

File: webapp/src/components/root/root.jsx [hunk 2]
```diff
@@ -2,11 +2,32 @@ import React from 'react';
...
-const Root = ({visible, close, theme}) => {
+import {menu} from '../../sub_menu';
+
+const Root = ({visible, close, theme, subMenuId}) => {
     if (!visible) {
         return null;
     }
 
+    let extraContent = '';
+    if (subMenuId) {
```
Comment (jespino) [inline webapp/src/components/root/root.jsx]: I think you are keeping this content after close, so if you open the root modal from any other place, you are seeing the "submenu" extra content here.

By the way, probably the extra content should be more descriptive about himself. You can see there "Third item" without any context, probably you should have something like `Element clicked in the menu: Third item`

File: webapp/i18n/es.json (hunk 1 above)
Comment (scottleedavis) [reply webapp/i18n/es.json]: Done

File: webapp/src/components/root/root.jsx (hunk 2 above)
Comment (scottleedavis) [reply webapp/src/components/root/root.jsx]: Great catch, I solved it by dispatching an empty subMenuId on normal Root modal trigger.  Also added the additional descriptive text.

**Comment by scottleedavis** [comment]
@jespino I simplified the data handling of the submenu structure in the root modal https://github.com/mattermost/mattermost-plugin-demo/pull/62/files#diff-d671a5593a8e810fd52c9f9f0361a48aR21-R25 Does this look ok to you?

**Comment by prapti** [comment]
Leaving the QA Review label on for now until all devs approve to make sure that this is tested again if there's any significant changes.

**Comment by jespino** [comment]
@scottleedavis yes, I think that addresses my concerns. Thanks! :)

File: webapp/src/actions.js [hunk 3]
```diff
@@ -4,6 +4,20 @@ import {id as pluginId} from './manifest';
 import {STATUS_CHANGE, OPEN_ROOT_MODAL, CLOSE_ROOT_MODAL} from './action_types';
 
 export const openRootModal = () => (dispatch) => {
+    dispatch({
```
Comment (marianunez) [inline webapp/src/actions.js]: Nit: could we have this call `openRootModalWithId('')` to avoid code duplication?

**Comment by scottleedavis** [comment]
Note all:  there will be some refactoring on the registry and the programmatic construction of the submenu as requested https://github.com/mattermost/mattermost-webapp/pull/3357#discussion_r319531033

**Comment by aaronrothschild** [comment]
@scottleedavis I'm not on the Repo so I can't remove the "PM Review" Label, sorry.

File: webapp/src/actions.js (hunk 3 above)
Comment (scottleedavis) [reply webapp/src/actions.js]: I agree, though I haven't found a clean way to do it such that dispatch is passed along.  I may be missing something.
```javascript
dispatch(openRootModalWithSubmenuName(''));   //doesn't work either
```

Thoughts on the cleanest way to accomplish this?

**Comment by scottleedavis** [comment]
no worries @aaronrothschild !   thank you.  👍

**Comment by scottleedavis** [comment]
Registry api is updated as part of the MM-16946 code review.  See plugin.jsx.

**Comment by scottleedavis** [comment]
@hanzei the build is failing with the known `unable to access 'https://git.apache.org/thrift.git/`.  Should I include the fix in this PR?

File: webapp/src/actions.js (hunk 3 above)
Comment (scottleedavis) [reply webapp/src/actions.js]: Opposed to leaving this in @marianunez ?  I am open to suggestions on changes, though don't have a better solution currently.

**Comment by hanzei** [comment]
That would be great @scottleedavis. You can copy the solution from https://github.com/mattermost/mattermost-server/pull/12146.

File: webapp/src/actions.js (hunk 3 above)
Comment (marianunez) [reply webapp/src/actions.js]: @scottleedavis I suggest keeping the original `openRootModal` and having the optional text parameter like:

```suggestion
    export const openRootModal = (subMenuText = '') => (dispatch) => {
    dispatch({
        type: SUBMENU,
        subMenu: subMenuText,
    });
    dispatch({
        type: OPEN_ROOT_MODAL,
    });
};
```

I tried it locally and seems to work fine for me. You can ping me in the community server if you want to talk it over.

Comment (scottleedavis) [reply webapp/src/actions.js]: This worked great, thank you @marianunez

File: webapp/src/actions.js [hunk 4]
```diff
@@ -1,9 +1,23 @@
...
+        type: SUBMENU,
+        subMenu: subMenuText,
+    });
+    dispatch({
+        type: OPEN_ROOT_MODAL,
+    });
+};
+
+export const openRootModalWithSubmenuName = (text) => (dispatch) => {
```
Comment (marianunez) [inline webapp/src/actions.js]: Now that we have `export const openRootModal = (subMenuText = '') => (dispatch)`,  `openRootModalWithSubmenuName` is no longer needed ?

Comment (scottleedavis) [reply webapp/src/actions.js]: Apologies for not having removed that.  Done

**Comment by marianunez** [comment]
I added the `Do Not Merge/Awaiting PR` label until we wrap up the related webapp changes and finalize when it will merge. We may need to update the required server version as well.

**Comment by hanzei** [comment]
Just to double this: This will require 5.18, correct?

**Comment by scottleedavis** [comment]
I believe so @hanzei .  
Updated the min_server_version in plugin.json

**Comment by hanzei** [comment]
This can be merged. Sorry for the delay, Scott.

**Pull request merged**